File: 08/08.c
#include <stdio.h>
#include <unistd.h>
#include <sys/types.h>
#include <signal.h>
#include <stdlib.h>

void handle_signal(int sig) {
    printf("Child Process: Received signal %d\n", sig);
    exit(0);
}

int main() {
    pid_t pid;
    
    // Create a child process using fork()
    pid = fork();
    
    if (pid == 0) {  // Child Process
        signal(SIGUSR1, handle_signal);  // Set signal handler
        printf("Child Process: Waiting for signal...\n");
        pause();  // Wait for signal
    } else if (pid > 0) {  // Parent Process
        sleep(2);  // Give the child process time to start and wait for signal
        printf("Parent Process: Sending SIGUSR1 to child\n");
        kill(pid, SIGUSR1);  // Send SIGUSR1 signal to the child
        wait(NULL);  // Wait for the child process to exit
    } else {
        perror("fork failed");
        exit(1);
    }

    return 0;
}

File: 08/08.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"syscall"
	"time"
)

func handleSignal(c chan os.Signal) {
	signal := <-c
	fmt.Printf("Child Process: Received signal %v\n", signal)
}

func main() {
	c := make(chan os.Signal, 1)

	// Create the child process using os/exec
	cmd := exec.Command(os.Args[0]) // Run the same program again to simulate a child process

	// Set the signal handler for the child
	go handleSignal(c)

	// Start the child process
	if err := cmd.Start(); err != nil {
		fmt.Println("Error starting child process:", err)
		return
	}

	// Parent process sends SIGUSR1 signal to the child after a delay
	time.Sleep(2 * time.Second) // Allow time for the child to start
	fmt.Println("Parent Process: Sending SIGUSR1 to child.")

	// Send signal to child process
	pid := cmd.Process.Pid
	syscall.Kill(pid, syscall.SIGUSR1) // Sending SIGUSR1 to the child

	// Wait for the child process to handle the signal
	time.Sleep(1 * time.Second)

	// Wait for the child process to finish
	cmd.Wait()
}
